### ISSUE TITLE
Extremely low accuracy

### ISSUE THREAD

**Issue by DhruvAwasthi** [ultralytics/yolov5#10532]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am trying to train a detection model to detect the regions in packaged food product images that contain the nutrients and ingredients. So basically there are two classes -`ingredients` and `nutrients`.  
  
The sample original image may looks like this:  

![2482-brazil-nuts-8oz-v4 5 1_web_back](https://user-images.githubusercontent.com/66522260/208471656-d4a68603-7c12-4310-90e8-4fd67b4697ad.jpg)  
  
  
For this, I labeled the training data and exported the `*.txt` files in `yolo` format. The sample .txt file looks like this:
```
0 0.366042 0.395625 0.632083 0.127917
1 0.367083 0.491042 0.634167 0.062917
```

I labeled around `250` images for this, and then after using augmentations the dataset size grew to over `6,000` images. In the training dataset, I have also included around 10% of random background images.

I organized the training data into the following format:
```
dataset/
    train/
        images/
        labels/
    val/
        images/
        labels/
    test/
        images/
        labels/
```

For training, I tried with `YOLOv5s`, `YOLOv5m`, and `YOLOv5m6`. I am training the model for around 300 epochs, using a patience level of `10`and using the `Adam` optimizer.   But due to early stopping the model training is stopping in between around `10 and 20 epochs`.

The `opt.yaml` file looks like this:
```
weights: yolov5m6.pt
cfg: ''
hyp:
  lr0: 0.01
  lrf: 0.01
  momentum: 0.937
  weight_decay: 0.0005
  warmup_epochs: 3.0
  warmup_momentum: 0.8
  warmup_bias_lr: 0.1
  box: 0.05
  cls: 0.5
  cls_pw: 1.0
  obj: 1.0
  obj_pw: 1.0
  iou_t: 0.2
  anchor_t: 4.0
  fl_gamma: 0.0
  hsv_h: 0.015
  hsv_s: 0.7
  hsv_v: 0.4
  degrees: 0.0
  translate: 0.1
  scale: 0.5
  shear: 0.0
  perspective: 0.0
  flipud: 0.0
  fliplr: 0.5
  mosaic: 1.0
  mixup: 0.0
  copy_paste: 0.0
epochs: 300
batch_size: 4
imgsz: 896
rect: false
resume: false
nosave: false
noval: false
noautoanchor: false
noplots: false
evolve: null
bucket: ''
cache: null
image_weights: false
device: ''
multi_scale: false
single_cls: false
optimizer: Adam
sync_bn: false
workers: 8
project: runs/train
name: exp
exist_ok: false
quad: false
cos_lr: false
label_smoothing: 0.0
patience: 10
freeze:
- 0
save_period: -1
seed: 0
local_rank: -1
entity: null
upload_dataset: false
bbox_interval: -1
artifact_alias: latest
save_dir: runs/train/exp
```
  

But even after all this, the confidence score that I am getting on the predictions on test images is too low, with the maximum being `0.016` and the lowest in the order of `0.0001xx`. These confidence scores are the same for the images that actually contain the classes that I want to predict, and for the images that do not contain these classes at all.    
  
  
Can you please tell me what you suspect is wrong here?

Thank you,
Dhruv

**Comment by glenn-jocher** [comment]
@DhruvAwasthi 👋 Hello! Thanks for asking about improving YOLOv5 🚀 training results. 

Most of the time good results can be obtained with no changes to the models or training settings, **provided your dataset is sufficiently large and well labelled**. If at first you don't get good results, there are steps you might be able to take to improve, but we always recommend users **first train with all default settings** before considering any changes. This helps establish a performance baseline and spot areas for improvement.

If you have questions about your training results **we recommend you provide the maximum amount of information possible** if you expect a helpful response, including results plots (train losses, val losses, P, R, mAP), PR curve, confusion matrix, training mosaics, test results and dataset statistics images such as labels.png. All of these are located in your `project/name` directory, typically `yolov5/runs/train/exp`.

We've put together a full guide for users looking to get the best results on their YOLOv5 trainings below.

## Dataset

- **Images per class.** ≥ 1500 images per class recommended
- **Instances per class.** ≥ 10000 instances (labeled objects) per class recommended
- **Image variety.** Must be representative of deployed environment. For real-world use cases we recommend images from different times of day, different seasons, different weather, different lighting, different angles, different sources (scraped online, collected locally, different cameras) etc.
- **Label consistency.** All instances of all classes in all images must be labelled. Partial labelling will not work.
- **Label accuracy.** Labels must closely enclose each object. No space should exist between an object and it's bounding box. No objects should be missing a label.
- **Label verification.** View `train_batch*.jpg` on train start to verify your labels appear correct, i.e. see [example](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#local-logging) mosaic.
- **Background images.** Background images are images with no objects that are added to a dataset to reduce False Positives (FP). We recommend about 0-10% background images to help reduce FPs (COCO has 1000 background images for reference, 1% of the total). No labels are required for background images.

<a href="https://arxiv.org/abs/1405.0312"><img width="800" src="https://user-images.githubusercontent.com/26833433/109398377-82b0ac00-78f1-11eb-9c76-cc7820669d0d.png" alt="COCO Analysis"></a>


## Model Selection

Larger models like YOLOv5x and [YOLOv5x6](https://github.com/ultralytics/yolov5/releases/tag/v5.0) will produce better results in nearly all cases, but have more parameters, require more CUDA memory to train, and are slower to run. For **mobile** deployments we recommend YOLOv5s/m, for **cloud** deployments we recommend YOLOv5l/x. See our README [table](https://github.com/ultralytics/yolov5#pretrained-checkpoints) for a full comparison of all models. 

<p align="center"><img width="700" alt="YOLOv5 Models" src="https://github.com/ultralytics/yolov5/releases/download/v1.0/model_comparison.png"></p>

- **Start from Pretrained weights.** Recommended for small to medium sized datasets (i.e. [VOC](https://github.com/ultralytics/yolov5/blob/master/data/VOC.yaml), [VisDrone](https://github.com/ultralytics/yolov5/blob/master/data/VisDrone.yaml), [GlobalWheat](https://github.com/ultralytics/yolov5/blob/master/data/GlobalWheat2020.yaml)). Pass the name of the model to the `--weights` argument. Models download automatically from the [latest YOLOv5 release](https://github.com/ultralytics/yolov5/releases).
```bash
python train.py --data custom.yaml --weights yolov5s.pt
                                             yolov5m.pt
                                             yolov5l.pt
                                             yolov5x.pt
                                             custom_pretrained.pt
```
- **Start from Scratch.** Recommended for large datasets (i.e. [COCO](https://github.com/ultralytics/yolov5/blob/master/data/coco.yaml), [Objects365](https://github.com/ultralytics/yolov5/blob/master/data/Objects365.yaml), [OIv6](https://storage.googleapis.com/openimages/web/index.html)). Pass the model architecture yaml you are interested in, along with an empty `--weights ''` argument:
```bash
python train.py --data custom.yaml --weights '' --cfg yolov5s.yaml
                                                      yolov5m.yaml
                                                      yolov5l.yaml
                                                      yolov5x.yaml
```


## Training Settings

Before modifying anything, **first train with default settings to establish a performance baseline**. A full list of train.py settings can be found in the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) argparser.

- **Epochs.** Start with 300 epochs. If this overfits early then you can reduce epochs. If overfitting does not occur after 300 epochs, train longer, i.e. 600, 1200 etc epochs.
- **Image size.** COCO trains at native resolution of `--img 640`, though due to the high amount of small objects in the dataset it can benefit from training at higher resolutions such as `--img 1280`. If there are many small objects then custom datasets will benefit from training at native or higher resolution. Best inference results are obtained at the same `--img` as the training was run at, i.e. if you train at `--img 1280` you should also test and detect at `--img 1280`.
- **Batch size.** Use the largest `--batch-size` that your hardware allows for. Small batch sizes produce poor batchnorm statistics and should be avoided.
- **Hyperparameters.** Default hyperparameters are in [hyp.scratch-low.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch-low.yaml). We recommend you train with default hyperparameters first before thinking of modifying any. In general, increasing augmentation hyperparameters will reduce and delay overfitting, allowing for longer trainings and higher final mAP. Reduction in loss component gain hyperparameters like `hyp['obj']` will help reduce overfitting in those specific loss components. For an automated method of optimizing these hyperparameters, see our [Hyperparameter Evolution Tutorial](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution).

## Further Reading

If you'd like to know more a good place to start is Karpathy's 'Recipe for Training Neural Networks', which has great ideas for training that apply broadly across all ML domains: http://karpathy.github.io/2019/04/25/recipe/

Good luck 🍀 and let us know if you have any other questions!

**Comment by huntracker** [comment]
@glenn-jocher I am trying to train a detection model to detect the coconut bunches,coconut stem,leaf,leaf stem,spathe in a custom coconut dataset i created. but the map is getting very low. i had tried with different epochs and batch size also changed the hyper parameters and tested but map is very low and not detecting all the classes while testing.
![result11](https://github.com/ultralytics/yolov5/assets/119437677/f843fa7d-fd6b-4e12-9bde-152b4a4cd4a4)
![yolo11](https://github.com/ultralytics/yolov5/assets/119437677/52bfb832-fbac-4e7a-9207-3c71605cba7e)

**Comment by glenn-jocher** [comment]
@huntracker 👋 Hello! It looks like you've put in a lot of effort creating a custom coconut dataset and training a detection model to recognize different coconut parts. 
The images you provided show the model's inference results but unfortunately it looks like the model is struggling to accurately detect the coconut parts.

It's great that you've experimented with different epochs, batch sizes, and hyperparameters, but there could be several other factors affecting the model's performance. Some suggestions to consider improving the mAP and class detection are:
1. **Increase the size of the dataset** by adding more diverse images and ensuring each class has enough representatives.
2. **Check the label quality** to ensure they closely enclose each object.
3. **Train for more epochs** if overfitting is not an issue.
4. **Check the dataset balance** to ensure the classes are represented equally.

If you haven't already, I'd recommend taking a look at the [YoloV5 Documentation](https://docs.ultralytics.com/) for detailed examples and best practices on creating custom datasets and training YOLOv5 models. 

Good luck 🍀 and let us know if you have any other questions!

**Comment by gulati-parth** [comment]
The default settings have nc (number of classes) listed as 80. Is there a specific reason for that/should we change it to our class number?

**Comment by glenn-jocher** [comment]
@gulati-parth yes, the default setting of `nc=80` corresponds to the number of classes in the COCO dataset, which YOLOv5 models are pre-trained on. When you are training on your custom dataset, you should definitely change `nc` to match the number of classes in your dataset. This is crucial for the model to correctly learn and predict your specific classes.

For example, if your dataset has 5 classes, you would set `nc=5` in your dataset's YAML file. This ensures the model's final layer is correctly configured to predict your dataset's classes.

Good luck with your training! 🚀 Let us know if you have any more questions.

**Comment by parth-gulati** [comment]
> @gulati-parth yes, the default setting of `nc=80` corresponds to the number of classes in the COCO dataset, which YOLOv5 models are pre-trained on. When you are training on your custom dataset, you should definitely change `nc` to match the number of classes in your dataset. This is crucial for the model to correctly learn and predict your specific classes.
> 
> For example, if your dataset has 5 classes, you would set `nc=5` in your dataset's YAML file. This ensures the model's final layer is correctly configured to predict your dataset's classes.
> 
> Good luck with your training! 🚀 Let us know if you have any more questions.

Thank you so much!!

**Comment by glenn-jocher** [comment]
@parth-gulati you're welcome! If you need further assistance or have more questions as you progress, feel free to reach out. Happy training! 🚀
